### PR TITLE
Fix potential thread leaking

### PR DESCRIPTION
**Pull request by efernandez**
fixes CORE-6291

This is done by setting pubThread_ to NULL whenever we join the thread,
so we only spawn a new thread if pubThread_ is NULL, instead of checking
the sub_ attribute.

### CODE REVIEW

File: pointgrey_camera_driver/src/nodelet.cpp [hunk 1]
```diff
@@ -59,22 +59,24 @@ class PointGreyCameraNodelet: public nodelet::Nodelet
...
   {
+    boost::mutex::scoped_lock scopedLock(connect_mutex_);
+
     if(pubThread_)
     {
       pubThread_->interrupt();
       pubThread_->join();
-    }
 
-    try
```
Comment (efernandez) [inline pointgrey_camera_driver/src/nodelet.cpp:68]: I don't think we need to stop + disconnect if the thread was already destroyed or never spawned.

File: pointgrey_camera_driver/src/nodelet.cpp [hunk 2]
```diff
@@ -158,34 +160,38 @@ class PointGreyCameraNodelet: public nodelet::Nodelet
...
-      scopedLock.lock();
-      sub_.shutdown();
-
-      try
-      {
-        NODELET_DEBUG("Stopping camera capture.");
-        pg_.stop();
-      }
-      catch(std::runtime_error& e)
+      if (pubThread_)
```
Comment (efernandez) [inline pointgrey_camera_driver/src/nodelet.cpp:163]: All the previous code is now indented inside this `if` and the only other change is `pubThread_ = NULL;`

File: pointgrey_camera_driver/src/nodelet.cpp [hunk 3]
```diff
@@ -158,34 +160,38 @@ class PointGreyCameraNodelet: public nodelet::Nodelet
...
+          NODELET_DEBUG("Disconnecting from camera.");
+          pg_.disconnect();
+        }
+        catch(std::runtime_error& e)
+        {
+          NODELET_ERROR("%s", e.what());
+        }
       }
     }
-    else if(!sub_)     // We need to connect
```
Comment (efernandez) [inline pointgrey_camera_driver/src/nodelet.cpp:188]: Now I check for `pubThread_` instead of `sub_`

**Comment by efernandez** [comment]
ping @afakihcpr @mikepurvis

**Review by afakihcpr — APPROVED** [review]
LGTM

**Comment by mikepurvis** [comment]
I'm fine with this if you want to merge it to test, but it's clear from our discussion that neither of us _really_ knows what it is that's going on in this code, or what caused the crashes we saw.

**Comment by efernandez** [comment]
Sorry, didn't wanted to click that button :wink:

**Comment by efernandez** [comment]
@mikepurvis I've just thought of another trace easier to understand that could definitely lead to leak a thread:

1. Client 1 subscribes **image_raw**, so [connectCb](https://github.com/clearpathrobotics/pointgrey_camera_driver/blob/master/pointgrey_camera_driver/src/nodelet.cpp#L154) gets called and it acquires the [connect_mutex_](https://github.com/clearpathrobotics/pointgrey_camera_driver/blob/master/pointgrey_camera_driver/src/nodelet.cpp#L157).
2. Since client 1 is connecting the condition `it_pub_.getNumSubscribers() == 0` fails.
3. Since it's the first client and the driver has just been started, `sub_` is `false`, so a new thread is created [here](https://github.com/clearpathrobotics/pointgrey_camera_driver/blob/master/pointgrey_camera_driver/src/nodelet.cpp#L191).
4. Now [devicePoll](https://github.com/clearpathrobotics/pointgrey_camera_driver/blob/master/pointgrey_camera_driver/src/nodelet.cpp#L295) starts running on another thread, but the camera takes time to initialize and we might spend some time connecting [here](https://github.com/clearpathrobotics/pointgrey_camera_driver/blob/master/pointgrey_camera_driver/src/nodelet.cpp#L368), which would communicate with the kernel and perform IO operations, so the thread would be taken out of the CPU.
5. Client 2 also subscribes **image_raw**, so [connectCb](https://github.com/clearpathrobotics/pointgrey_camera_driver/blob/master/pointgrey_camera_driver/src/nodelet.cpp#L154) gets called and it acquires the [connect_mutex_](https://github.com/clearpathrobotics/pointgrey_camera_driver/blob/master/pointgrey_camera_driver/src/nodelet.cpp#L157).
6. Now, even if client 1 receives more CPU time and finishes the connection with the camera, it CANNOT acquire the mutex, and initialize the subscriber [here](https://github.com/clearpathrobotics/pointgrey_camera_driver/blob/master/pointgrey_camera_driver/src/nodelet.cpp#L391).
7. Since client 2 is connecting the condition `it_pub_.getNumSubscribers() == 0` fails.
8. Since client 1 couldn't create the subscriber yet, `sub_` is still `false`, so another thread is created [here](https://github.com/clearpathrobotics/pointgrey_camera_driver/blob/master/pointgrey_camera_driver/src/nodelet.cpp#L191) and the previous one is leaked.

**Comment by mikepurvis** [comment]
Okay, yes, I definitely agree with that possibility. As discussed in person, it's interesting that we don't see this as a crash all the time, though (or conversely, that we saw it ever, given that we have only a single subscriber to `image_raw`).

TBH, I'd be perfectly fine with ripping out all of this connection logic and simplifying the driver down to just running and publishing all the time.

**Comment by efernandez** [comment]
I'm reluctant to have the camera streaming images all the time as these cameras usually get very hot and they consume a considerable amount of power. IMHO, some people might actually prefer the lazy publisher (including the camera streaming) behaviour.

**Comment by mikepurvis** [comment]
Fair enough, but in our use case, the camera is definitely streaming all the time, isn't it? Or is the intent that on a bidirectional robot, we would unsubscribe from the rear-facing cams when traveling in the other direction, for example?

**Comment by servos** [comment]
There was some discussion of enabling/disabling the cameras based on direction of travel, however I think that we would still need to keep the cameras on as we would not want to have the delay caused by connecting to the cameras (but this has not been explored currently).

**Comment by efernandez** [comment]
My comment was on regard of other/external users of this driver, although this is actually a fork.

**Comment by mikepurvis** [comment]
Yeah for sure. I would like to examine our deltas and get everything merged upstream, though— I'm the maintainer of the upstream driver, and I'd feel a lot more confident in making releases of it if it was the same code we're testing in house here.

**Comment by efernandez** [comment]
@mikepurvis This has been tested with the Point Grey camera and we couldn't make it crash.

I'm going to merge it... anything comment before I do?

**Comment by mikepurvis** [comment]
Sounds good.

**Pull request merged**